### ISSUE TITLE
RabbitmqCluster lacks status.observedGeneration

### ISSUE THREAD

**Issue by n3wscott** [rabbitmq/cluster-operator#311]
## Describe the bug
Status does not contain an `observedGeneration`. See: https://github.com/rabbitmq/cluster-operator/blob/main/api/v1beta1/rabbitmqcluster_types.go#L282-L289

## To Reproduce
1. Look at status.
2. Be not sure if the spec has been propagated to status.

**Expected behavior**

Resources in Kubernetes should propagate their `ObservedGeneration` when reconciliation of the spec has been realized for a given generation.

## Additional context
Knative recommends the following minimum shape for a status: https://github.com/knative/pkg/blob/master/apis/duck/v1/status_types.go#L38-L55

**Comment by Zerpet** [comment]
Hi Scott 👋 
Our initial design of the `Status` field was based on the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties). The last paragraph in this section mentions briefly the `observedGeneration` field, however it doesn't mention if this field may, should or must be included in the `Status` field, so we didn't consider this field in our first iteration of the status.

In our initial design, the method to know if the `Spec` has been propagated was via the status condition `ReconcileSuccess`. If this condition is `True`, then our Operator did the reconcilation successfully.

What would be the benefits of implementing `ObservedGeneration` ?
In other words, how would you use the `ObservedGeneration` in the `RabbitmqCluster` resource in your use case?

**Comment by RichiCoder1** [comment]
At least from my understanding, ObservedGeneration has value in the case:
1. A resource is update whilst the controller is down and so the generation != observedGeneration, so the controller knows it needs to reconcile
2. that a third-party can act upon status updates only when the status is in-sync with the latest generation

That's my very, very basic understanding though.

**Comment by Zerpet** [comment]
> A resource is update whilst the controller is down and so the generation != observedGeneration, so the controller knows it needs to reconcile

This makes a lot of sense and actually is not very hard to implement, I will bring it up to the team, thanks!

Recording context here:

Kubernetes seems to assign a `generation` field to the metadata of certain objects. It does this, among others, for `Deployment` and `StatefulSet`. It is doing the same for `RabbitmqCluster` resource. The idea would be to set `status.ObservedGeneration = metadata.generation` at the end of the reconcile loop, once the reconciliation completed without errors. As an optimisation, we can also have a check at the beginning of the reconcile function to return success immediately if `observedGeneration == generation`.

The only caveat or "different" case I can think of is enabling/disabling plugins.

**Comment by Zerpet** [comment]
I've made a quick test around this, these are my findings:

- Setting `status.ObservedGeneration` is quite easy, just a line at the end of the `Reconcile()` function
- We can't quite assume that if `generation == observedGeneration`, we don't have to reconcile because:
  - We use the reconcile loop to create/update child resources e.g. Service, StatefulSet, etc.
  - When an event happens in a child resource, the generation of `RabbitmqCluster` does not change.
- If the Operator goes down, e.g. pod deleted, the new Operator that comes up receives requests to reconcile all existing `RabbitmqClusters`.

I looked at the StatefulSet controller code [here](https://github.com/kubernetes/kubernetes/blob/fb02a59a6a10623b971107231b3c823fed6c43a8/pkg/controller/statefulset/stateful_set.go#L385-L455) and [here](https://github.com/kubernetes/kubernetes/blob/fb02a59a6a10623b971107231b3c823fed6c43a8/pkg/controller/statefulset/stateful_set_control.go#L271-L293) and they do not leverage observed generation to determine if the StatefulSet is up to date. This field seems have an informative purpose and no relevance in the behaviour of the controller itself.

We can still prioritise this issue regardless.
